### PR TITLE
fix regression where typealiases are not executed in their original context

### LINKED ISSUE #823: Regression: incorrect scope in T of typealias of Listing<T>/Mapping<K, T>
Given:

```pkl
// mod1.pkl
const local lastName = "Smith"

typealias Smiths = Listing<String(endsWith(lastName))>
```

This breaks:

```pkl
import "mod1.pkl"

res: mod1.Smiths = new { "Lily Smith" }
```

Error:

```
–– Pkl Error ––
Cannot find property `lastName`.

3 | typealias Smiths = Listing<String(endsWith(lastName))>
                                               ^^^^^^^^
```

### PR DESCRIPTION
**Pull request by stackoverflow**
Closes #823

### CODE REVIEW

File: pkl-core/src/main/java/org/pkl/core/ast/member/ListingOrMappingTypeCastNode.java [hunk 1]
```diff
@@ -54,6 +73,20 @@ public SourceSection getSourceSection() {
...
+      VmUtils.setOwner(frame, owner);
+      VmUtils.setReceiver(frame, receiver);
+      return executeInternal(frame);
+    } finally {
+      VmUtils.setOwner(frame, prevOwner);
+      VmUtils.setReceiver(frame, prevReceiver);
+    }
+  }
+
+  private Object executeInternal(VirtualFrame frame) {
```
Comment (holzensp) [inline pkl-core/src/main/java/org/pkl/core/ast/member/ListingOrMappingTypeCastNode.java:89]: mild preference to just inline, myself
```suggestion
      return typeNode.execute(frame, frame.getArguments()[2]);
    } catch (VmTypeMismatchException e) {
      CompilerDirectives.transferToInterpreter();
      throw e.toVmException();
    } finally {
      VmUtils.setOwner(frame, prevOwner);
      VmUtils.setReceiver(frame, prevReceiver);
    }
  }
```

(GH suggestion can't span out of the diff, alas)

File: pkl-core/src/main/java/org/pkl/core/ast/type/TypeNode.java [hunk 2]
```diff
@@ -1342,6 +1373,14 @@ public Object executeEagerly(VirtualFrame frame, Object value) {
       throw typeMismatch(value, BaseModule.getMapClass());
     }
 
+    @Override
+    public Object executeInDefinitionContext(VirtualFrame frame, Object value) {
+      if (value instanceof VmMap vmMap) {
+        return evalInDefinitionContext(frame, vmMap);
+      }
+      throw typeMismatch(value, BaseModule.getMapClass());
```
Comment (holzensp) [inline pkl-core/src/main/java/org/pkl/core/ast/type/TypeNode.java:1381]: For consistency and to save on single use (albeit `private`) helpers; why not follow the same pattern as in `VmList`?

File: pkl-core/src/main/java/org/pkl/core/ast/type/TypeNode.java [hunk 3]
```diff
@@ -1427,7 +1483,15 @@ public Object execute(VirtualFrame frame, Object value) {
...
-      return doTypeCast(frame, vmListing);
+      return doTypeCast(frame, vmListing, false);
+    }
+
+    @Override
+    public Object executeInDefinitionContext(VirtualFrame frame, Object value) {
+      if (!(value instanceof VmListing vmListing)) {
+        throw typeMismatch(value, BaseModule.getListingClass());
+      }
+      return doTypeCast(frame, vmListing, true);
```
Comment (holzensp) [inline pkl-core/src/main/java/org/pkl/core/ast/type/TypeNode.java:1494]: ^^ same re:`VmList`

File: pkl-core/src/main/java/org/pkl/core/ast/type/TypeNode.java [hunk 4]
```diff
@@ -1560,15 +1634,26 @@ public TypeNode getValueTypeNode() {
...
+                  valueTypeNode,
+                  getRootNode().getName(),
+                  frame);
+        } else {
+          listingOrMappingTypeCastNode =
+              new ListingOrMappingTypeCastNode(
+                  VmLanguage.get(this),
+                  getRootNode().getFrameDescriptor(),
+                  valueTypeNode,
+                  getRootNode().getName());
```
Comment (holzensp) [inline pkl-core/src/main/java/org/pkl/core/ast/type/TypeNode.java:1655]: It's this kind of repetition that would almost make me suggest to just make passing a `null` part of the API of `ListingOrMappingTypeCastNode`

**Review by holzensp — APPROVED** [review]
Solves a problem, but less than ideally pretty (I think we agree). We may want to do a clean-up pass on how we do frame management.

File: pkl-core/src/main/java/org/pkl/core/ast/member/ListingOrMappingTypeCastNode.java (hunk 1 above)
Comment (stackoverflow) [reply pkl-core/src/main/java/org/pkl/core/ast/member/ListingOrMappingTypeCastNode.java:89]: Inlining would make it slower, right? I'd always have to get and reset the owner/receiver in the frame, while right now it's short-circuiting for the 99% case where it doesn't come from a typealias.

File: pkl-core/src/main/java/org/pkl/core/ast/type/TypeNode.java (hunk 2 above)
Comment (stackoverflow) [reply pkl-core/src/main/java/org/pkl/core/ast/type/TypeNode.java:1381]: I did this for consistency with `execute` and `executeEager`. I don't think it makes sense to only inline it here. Either we remove `eval` and `evalEager` or we leave all three.

File: pkl-core/src/main/java/org/pkl/core/ast/type/TypeNode.java (hunk 4 above)
Comment (stackoverflow) [reply pkl-core/src/main/java/org/pkl/core/ast/type/TypeNode.java]: Improved.

File: pkl-core/src/main/java/org/pkl/core/ast/member/ListingOrMappingTypeCastNode.java (hunk 1 above)
Comment (holzensp) [reply pkl-core/src/main/java/org/pkl/core/ast/member/ListingOrMappingTypeCastNode.java:85]: You are entirely right... I had missed that top line somehow

**Comment by odenix** [comment]
I implemented a simpler fix that passes your tests for my "lazy-check" branch: https://github.com/apple/pkl/pull/789/commits/bc10eba3b27336caba257333e928da1a3c6d8c81

PS: As far as I know, having multiple execute methods with the same signature isn't a common Truffle idiom and can cause problems with `@Specialization`. Instead of having `TypeNode.execute` and `TypeNode.executeEagerly`, it might be better to have `ListingTypeNode` and `EagerListingTypeNode` (same for mappings).

**Comment by odenix** [comment]
It would be great if we could prioritize https://github.com/apple/pkl/pull/789 to avoid parallel development in this part of the codebase.

**Comment by stackoverflow** [comment]
@odenix I'll review you PR and then, hopefully, we can close this one.

**Pull request closed without merging**